### PR TITLE
RasterTorchgeo Update

### PR DESCRIPTION
**Pull request by tlarcher**
# Reference Issues/PRs
This pull requests merges to `main` changes brought by Issue #8 , Issue #14 , PR #13 , PR #15 and PR #16.

# What does this implement/fix? Explain your changes.
## Major improvements

- Added a new class `RasterTorchGeoDataset` inheriting torchgeo's `RasterDataset` class.

This class allows to provide both patches from voluminous, multi-modality, rasters containing geospatial information, and associated labels for 3 classification tasks : binary, multi-class and multi-label. The class benefits from torchgeo's tiles mapping and overlapping prevention while offering the possibility to query both the data and the associated labels that are being fed to PyTorch dataloaders. For that purpose, a custom sampler has been written.

Another advantage of this class is that it allows the user to query the dataset with any type of coordinates, and any unit amongst **[meter, pixel, crs]**, even if the dataset's CRS or CRS' unit doesn't match the user's query.

New examples have been generated, using this class, with automatically downloaded Sentinel-2 data samples (one 4-bands tile over the city of Montpellier, France) from Microsoft Planetary Computer.

Unit tests have been written and passed for and end-to-end validation, as well as for every method used by the class.

- Added checkpoint training resume strategy by default.

The checkpoint weight file path is customizable through the config. file.

- Updated config files

Configuration files have been enriched in accordance to the new features delivered by this PR. New arguments have been added to specify the type of classification task of the associated experiment, the chosen metrics, or the user's query parameters passed on to PyTorch dataloaders.

## Minor improvements
- Full code documentation support added for `RasterTorchGeoDataset` and new `utils` methods.
- Created new READMEs for `RasterTorchGeoDataset` examples and updated existing READMEs to match config files updates.

# Any other comments?
Enriching the examples READMEs to make them more straightforward and complete would be good.

### CODE REVIEW

File: malpolon/models/standard_prediction_systems.py [hunk 1]
```diff
@@ -59,6 +59,14 @@ def __init__(
     def forward(self, x: Any) -> Any:
         return self.model(x)
 
+    def _cast_type_to_loss(self, y):
```
Comment (tlarcher) [inline malpolon/models/standard_prediction_systems.py:62]: Source ? Add confusion matrices in a resources folder of the tests directory

File: malpolon/tests/extract_sample_from_raster.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
+"""This script helps extracting small rasters from big rasters.
+
+This script is intended for testing purposes.
+
+Author: Theo Larcher <theo.larcher@inria.fr>
+"""
+
+import rasterio
+import rioxarray
```
Comment (tlarcher) [inline malpolon/tests/extract_sample_from_raster.py:9]: Unused
```suggestion

```

File: malpolon/tests/extract_sample_from_raster.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    with rasterio.open(full_fp) as src:
+        full = src.read()
+        full_crs = src.crs
+    with rasterio.open(sample_fp) as src:
+        sample = src.read()
+        sample_crs = src.crs
+    xy_offset = (full.shape[2] // 2 - size[0] // 2, full.shape[1] // 2 - size[1] // 2) if xy_offset is None else xy_offset
+    assert full_crs.to_epsg() == sample_crs.to_epsg()
+    assert full_crs == sample_crs
+    # assert np.array_equal(full[0, xy_offset[0]:xy_offset[0]+size[0], xy_offset[1]:xy_offset[1]+size[1]], sample[0])
```
Comment (tlarcher) [inline malpolon/tests/extract_sample_from_raster.py:83]: ? Probably best to use `np.testing.assert_allclose`
```suggestion
    # np.testing.assert_allclose(full[0, xy_offset[0]:xy_offset[0]+size[0], xy_offset[1]:xy_offset[1]+size[1]], sample[0])
```
\+ uncomment ?

File: malpolon/tests/test_environmental_raster.py [hunk 4]
```diff
@@ -2,11 +2,11 @@
 
 import numpy as np
 import pytest
+import torch
 
 from malpolon.data.environmental_raster import PatchExtractor
 
-
-DATA_PATH = Path("/home/lorieul/dev/research/GeoLifeCLEF22/data")
+DATA_PATH = Path("malpolon/tests/")
```
Comment (tlarcher) [inline malpolon/tests/test_environmental_raster.py:9]: It seems the DATA_PATH is wrong or obsolete and should point to `malpolon.tests.data`
```suggestion
DATA_PATH = Path("malpolon/tests/data")
```
Also, the subdirectory `rasters` is missing so the tests will fail on missing files.

File: malpolon/data/datasets/torchgeo_datasets.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+
+class RasterTorchGeoDataset(RasterDataset):
+    """Generic torchgeo based raster datasets.
+
+    Datasets based on this class return patches from raster files and
+    can be queried by either a torchgeo BoundingBox object, a tuple of
+    coordinates in the dataset's CRS or a dictionary specifying
+    coordinates and the wanted CRS. Additionally one can specify the
+    desired size and units of the wanted patch.
```
Comment (tlarcher) [inline malpolon/data/datasets/torchgeo_datasets.py:41]:
```suggestion
    desired size and units of the wanted patch even if they don't match the dataset's.
    For instance, on can query a lon/lat dataset in EPSG:4326 with a 200x200 meters bounding box.
```

File: malpolon/data/datasets/torchgeo_datasets.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        self.task = task
+        self.binary_positive_classes = set(binary_positive_classes)
+        self.transforms_data = transforms_data
+
+        df = self._load_observation_data(Path(root), labels_name, split)
+        self.observation_ids = df.index
+        self.coordinates = df[["longitude", "latitude"]].values
+        self.targets = df["species_id"].values
+
+    def _load_observation_data(
```
Comment (tlarcher) [inline malpolon/data/datasets/torchgeo_datasets.py:107]: Add default values of variables in docstring

File: malpolon/data/datasets/torchgeo_datasets.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+        units : str, optional
+            The coordinates' unit system, must have a value in ['pixel', 'crs'].
+            The size of the bbox will adapt to the unit. If 'pixel' is
+            selected, the bbox size will be multiplied by the dataset
+            resolution. Selecting 'crs' will not modify the bbox size. In that
+            case the returned bbox will be of size:
+            (size[0], size[1]) <metric_of_the_dataset (usually meters)>.
+            Defaults to 'crs'.
+        crs : Union[int, str]
+            CRS of the point's lon/lat coordinates, by default None.
```
Comment (tlarcher) [inline malpolon/data/datasets/torchgeo_datasets.py:234]:
```suggestion
            CRS of the point's lon/lat coordinates, by default 'self'.
```

File: malpolon/data/datasets/torchgeo_datasets.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+
+        Parameters
+        ----------
+        lon : Union[int, float]
+            longitude
+        lat : Union[int, float]
+            latitude
+        size : Union[tuple, int], optional
+            Patch size, by default None. If passed as an int, the patch will be
+            square. If passed as a tuple (width, height), can be rectangular.
```
Comment (tlarcher) [inline malpolon/data/datasets/torchgeo_datasets.py:224]:
```suggestion
            square. If passed as a tuple (width, height), can be rectangular,
            by default None
```

**Review by tlarcher — COMMENTED** [review]
Several comments to adress

File: examples/micro_geolifeclef/cnn_on_rgb_nir_patches.py [hunk 9]
```diff
@@ -115,19 +122,54 @@ def __call__(self, old_layer, new_layer):
...
+            by default True
+        """
+        if hparams_preprocess:
+            task = task.split('classification_')[1]
+            metrics = omegaconf.OmegaConf.to_container(metrics)
+            for k, v in metrics.items():
+                if 'callable' in v:
+                    metrics[k]['callable'] = eval(v['callable'])
+                else:
+                    metrics[k]['callable'] = FMETRICS_CALLABLES[k]
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply examples/micro_geolifeclef/cnn_on_rgb_nir_patches.py]: ok

File: examples/micro_geolifeclef/cnn_on_rgb_patches.py [hunk 10]
```diff
@@ -89,17 +90,53 @@ def get_dataset(self, split, transform, **kwargs):
...
+            by default True
+        """
+        if hparams_preprocess:
+            task = task.split('classification_')[1]
+            metrics = omegaconf.OmegaConf.to_container(metrics)
+            for k, v in metrics.items():
+                if 'callable' in v:
+                    metrics[k]['callable'] = eval(v['callable'])
+                else:
+                    metrics[k]['callable'] = FMETRICS_CALLABLES[k]
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply examples/micro_geolifeclef/cnn_on_rgb_patches.py]: ok

File: examples/sentinel2_raster_torchgeo/cnn_on_rgbnir_torchgeo.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+            self.download_data_sample()
+
+    def download_data_sample(self):
+        """Download 4 Sentinel-2A tiles from MPC.
+
+        This method is useful to quickly download a sample of
+        Sentinel-2A tiles via Microsoft Planetary Computer (MPC).
+        The referenced of the tile downloaded are specified by the
+        `tile_id` and `timestamp` variables. Tiles are not downloaded
+        if they already have are found locally.
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply examples/sentinel2_raster_torchgeo/cnn_on_rgbnir_torchgeo.py]: ok

File: examples/sentinel2_raster_torchgeo/cnn_on_rgbnir_torchgeo.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+            by default True
+        """
+        if hparams_preprocess:
+            task = task.split('classification_')[1]
+            metrics = omegaconf.OmegaConf.to_container(metrics)
+            for k, v in metrics.items():
+                if 'callable' in v:
+                    metrics[k]['callable'] = eval(v['callable'])
+                else:
+                    metrics[k]['callable'] = FMETRICS_CALLABLES[k]
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply examples/sentinel2_raster_torchgeo/cnn_on_rgbnir_torchgeo.py]: ok

File: examples/sentinel2_raster_torchgeo/transforms.py
[the remark this answers is not in the document]
Comment (tlarcher) [reply examples/sentinel2_raster_torchgeo/transforms.py:1]: ok

File: malpolon/data/data_module.py [hunk 13]
```diff
@@ -5,6 +5,7 @@
 
 import pytorch_lightning as pl
 from torch.utils.data import DataLoader
+from malpolon.data.datasets.torchgeo_datasets import Sentinel2GeoSampler
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply malpolon/data/data_module.py:8]: ok

File: malpolon/data/environmental_raster.py [hunk 14]
```diff
@@ -1,12 +1,13 @@
 from __future__ import annotations
 import warnings
 from pathlib import Path
-from typing import Any, Optional, Union, TYPE_CHECKING
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, Union, TYPE_CHECKING
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply malpolon/data/environmental_raster.py:4]: ok

File: malpolon/data/utils.py [hunk 15]
```diff
@@ -1,7 +1,112 @@
 """This file compiles useful functions related to data and file handling."""
+from __future__ import annotations
 
 import os
 import re
+import numpy as np
+from typing import TYPE_CHECKING, Iterable, Union
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply malpolon/data/utils.py:7]: ok

File: malpolon/tests/extract_sample_from_raster.py (hunk 2 above)
Comment (tlarcher) [reply malpolon/tests/extract_sample_from_raster.py:9]: ok

File: malpolon/tests/test_environmental_raster.py [hunk 16]
```diff
@@ -2,11 +2,11 @@
 
 import numpy as np
 import pytest
+import torch
```
[the remark this answers is not in the document]
Comment (tlarcher) [reply malpolon/tests/test_environmental_raster.py:5]: ok

File: malpolon/tests/test_environmental_raster.py (hunk 4 above)
Comment (tlarcher) [reply malpolon/tests/test_environmental_raster.py:8]: ok. Added subdirectory `bioclimatic_rasters` with 3 rasters (18,2MB)

File: malpolon/data/datasets/torchgeo_datasets.py (hunk 5 above)
Comment (tlarcher) [reply malpolon/data/datasets/torchgeo_datasets.py:41]: ok

File: malpolon/data/datasets/torchgeo_datasets.py (hunk 6 above)
Comment (tlarcher) [reply malpolon/data/datasets/torchgeo_datasets.py:107]: ok

File: malpolon/data/datasets/torchgeo_datasets.py (hunk 8 above)
Comment (tlarcher) [reply malpolon/data/datasets/torchgeo_datasets.py:224]: ok

File: malpolon/tests/extract_sample_from_raster.py (hunk 3 above)
Comment (tlarcher) [reply malpolon/tests/extract_sample_from_raster.py]: ok

File: malpolon/data/utils.py
[the remark this answers is not in the document]
Comment (tlarcher) [reply malpolon/data/utils.py:1]: ok

File: malpolon/models/standard_prediction_systems.py (hunk 1 above)
Comment (tlarcher) [reply malpolon/models/standard_prediction_systems.py:62]: ok

**Comment by tlarcher** [comment]
lgtm
